lib/galaxy/managers/jobs.py

```diff
@@ -1908,3 +1940,107 @@ def get_jobs_to_check_at_startup(session: galaxy_scoped_session, track_jobs_in_d
 def get_job(session: galaxy_scoped_session, *where_clauses):
     stmt = select(Job).where(*where_clauses).limit(1)
     return session.scalars(stmt).first()
+
+
+@dataclass
+class DereferencedDatasetPair:
+    hda: model.HistoryDatasetAssociation
+    request: DataRequestUri
+
+
+class JobSubmitter:
+    def __init__(
+        self,
+        history_manager: HistoryManager,
+        user_manager: UserManager,
+        hda_manager: HDAManager,
+        app: MinimalManagerApp,
+    ):
+        self.history_manager = history_manager
+        self.user_manager = user_manager
+        self.hda_manager = hda_manager
+        self.app = app
+
+    def materialize_request_for(
+        self, trans: WorkRequestContext, hda: model.HistoryDatasetAssociation
+    ) -> MaterializeDatasetInstanceTaskRequest:
+        return MaterializeDatasetInstanceTaskRequest(
+            user=trans.async_request_user,
+            history_id=trans.history.id,
+            source="hda",
+            content=hda.id,
+        )
+
+    def dereference(
+        self, trans: WorkRequestContext, tool: Tool, request: QueueJobs, tool_request: ToolRequest
+    ) -> Tuple[RequestInternalDereferencedToolState, list[DereferencedDatasetPair]]:
+        new_hdas: list[DereferencedDatasetPair] = []
+
+        def dereference_callback(data_request: DataRequestUri) -> DataRequestInternalHda:
+            # a deferred dataset corresponding to request
+            history = tool_request.history
+            if not history:
+                raise InconsistentDatabase("Tool request has no history associated")
+
+            hda = dereference_input(trans, data_request, history)
+            if not isinstance(hda, model.HistoryDatasetAssociation):
+                raise RequestParameterInvalidException("Input dataset is not a history dataset association")
+
+            new_hdas.append(DereferencedDatasetPair(hda, data_request))
+            return DataRequestInternalHda(id=hda.id, src="hda")
+
+        tool_state = RequestInternalToolState(tool_request.request)
+        return dereference(tool_state, tool, dereference_callback), new_hdas
+
+    def queue_jobs(self, tool: Tool, request: QueueJobs) -> None:
+        tool_request: ToolRequest = self._tool_request(request.tool_request_id)
+        sa_session = self.app.model.context
+        try:
+            request_context = self._context(tool_request, request)
+            target_history = request_context.history
+            use_cached_jobs = request.use_cached_jobs
+            rerun_remap_job_id = request.rerun_remap_job_id
+            tool_state: RequestInternalDereferencedToolState
+            new_hdas: list[DereferencedDatasetPair]
```

Comment:
These type hints seem unnecessary ?

Reply:
necessary is irrelevant - they make the code more grokable - but I will drop them for you begrudgingly.